File: thread_performance_test/Java/MatrixMultiplication.java
import java.util.*;
import java.time.*;
import java.math.BigDecimal;


class MatrixMultiplication extends Thread {
    static int MAX = 1000;
    static int MAX_THREAD = 4;

    static int matA[][];
    static int matB[][];
    static int matC[][];
    static int step_i = 0;


    public void run(){
        int core = step_i++; 

	// Each thread computes 1/4th of matrix multiplication 
	for (int i = core * MAX / 5; i < (core + 1) * MAX / 5; i++) 
		for (int j = 0; j < MAX; j++) 
			for (int k = 0; k < MAX; k++) 
				matC[i][j] += matA[i][k] * matB[k][j]; 

              
    }
    public static void main(String [] args){

        matA = new int[MAX][MAX];
        matB = new int[MAX][MAX];
        matC = new int[MAX][MAX];

        Random rand = new Random();

        for (int i = 0; i < MAX; i++) { 
            for (int j = 0; j < MAX; j++) { 
                matA[i][j] = rand.nextInt(10); 
                matB[i][j] = rand.nextInt(10); 
            } 
        } 

        System.out.println("Matrix A");
        for (int i = 0; i < MAX; i++) { 
            for (int j = 0; j < MAX; j++){
                System.out.print(matA[i][j]+" ");
            }
            System.out.println();
        }

        System.out.println("Matrix B");
        for (int i = 0; i < MAX; i++) { 
            for (int j = 0; j < MAX; j++){
                System.out.print(matB[i][j]+" ");
            }
            System.out.println();
        }
        long startTime = System.nanoTime();   

        try{
            List<Thread> threads = new ArrayList<>(MAX_THREAD);
      
              for (int x = 0; x < MAX_THREAD; x++) {
                  Thread t = new Thread(new MatrixMultiplication());
                  t.start();
      
      
                 threads.add(t);
                
                    
              // System.out.println("Get Result: " + t.getResult());
              
              }
              for (Thread t : threads) {
                  t.join();
               }
              
          }catch (Exception e) {
              e.printStackTrace();
          } 

          long endTime = System.nanoTime(); 
          
          long elapsedTime = endTime - startTime;


          double seconds = (double)elapsedTime / 1000000000;


          System.out.println();
          System.out.println("Multiplication of A and B");
          for (int i = 0; i < MAX; i++) { 
            for (int j = 0; j < MAX; j++){
                System.out.print(matC[i][j]+" ");
            }
            System.out.println();
        }
        System.out.println("Time Taken: "+BigDecimal.valueOf(seconds).toPlainString());

    }



}

File: thread_performance_test/Python/Matrix_Multiplication.py

import threading
import random
import time


THREAD_MAX = 4
MAX = 1000
step_i = 0
matA = []
matB = []
matC = []
  
def multi(matA,matB,matC):
    global step_i
    core = step_i
    step_i = step_i+1
    start = int(core * MAX / 5)
    end = int((core + 1) * MAX / 5)

    for i in range(start,end):
        for j in range(0,MAX):
            for k in range(0,MAX):
                matC[i][j] += matA[i][k] * matB[k][j]


    return                


if __name__ == '__main__': 
# Test array 

    
  
# Function call 
    
    # starting threads
    for r in range(0, MAX):
        matA.append([random.randint(0, 2147483647)%10 for c in range(0, MAX)])

    for r in range(0, MAX):
        matB.append([random.randint(0, 2147483647)%10 for c in range(0, MAX)])


    for r in range(0, MAX):
       matC.append([0 for c in range(0, MAX)])    

    
    
    start_time = time.time()

    Threads = []
   
    for q in range(THREAD_MAX):
        t = threading.Thread(target=multi, args=(matA,matB,matC))
        Threads.append(t)
        t.start()

    for q in range(THREAD_MAX):    
        Threads[q].join()

    print()
    print("Matrix A")

    for i in range(0,MAX):
        for j in range(0,MAX):
            print(matA[i][j], end =" ")
        print()    

    print()
    print("Matrix B")

    for i in range(0,MAX):
        for j in range(0,MAX):
            print(matB[i][j], end =" ")
        print()    
                      

 
    print()
    print("Multiplication of A and B")

    for i in range(0,MAX):
        for j in range(0,MAX):
            print(matC[i][j], end =" ")
        print()

    print("Time taken:", time.time() - start_time, "seconds.")
        
           
